stdlib/iter.mli

```diff
@@ -0,0 +1,42 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                 Simon Cruanes                                          *)
+(*                                                                        *)
+(*   Copyright 2001 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* Module [Iter]: functional iterators *)
+
+(** @since NEXT_RELEASE *)
+
+type ('a, 's) step =
+  | Done
+  | Skip of 's
+  | Yield of 'a * 's
+
+type +_ t =
+  | Sequence : 's * ('s -> ('a,'s) step) -> 'a t
+
+val empty : 'a t
+
+val return : 'a -> 'a t
+
+val map : ('a -> 'b) -> 'a t -> 'b t
+
+val filter : ('a -> bool) -> 'a t -> 'a t
+
+val filter_map : ('a -> 'b option) -> 'a t -> 'b t
+
+val flat_map : ('a -> 'b t) -> 'a t -> 'b t
+
+val fold_left : ('a -> 'b -> 'a) -> 'a -> 'b t -> 'a
```

Comment:
@Maelan Ideed, nesting `left_fold` works as expected. I'm aware of the `sum` case, though I've considered that "coincidental" currying more than something which can be used systematically.  Your `|>` would chain with a different meaning, namely stepwise brewing down a nested container to(wards) a scalar.  I'm not convinced though, but I may have a bias:  My preferred understanding of folds is in terms of compositions of automorphisms (`fold f = f x₁ ∘ ... ∘ f xₙ`). I suspect this view gives a good correspondence between the layout of the program and what's going on (nested expressions correspond to processing nested containers), but there may be more nice properties of `fold_left` which I'm missing.  In any case this discussion could become very long, so

@c-cube If you're not convinced, `fold_left` is okay too, and that would leave room to add a `fold` at a later point after a longer discussion (read: bikeshedding session).